Subject: allow handling of x86-64 crash dumps on i386
From: http://xenbits.xensource.com/xen-unstable.hg (tip 14960)
Acked-by: jbeulich@novell.com

From: Ian Campbell <ian.campbell@xensource.com>

The specific case I am encountering is kdump under Xen with a 64 bit
hypervisor and 32 bit kernel/userspace.  The dump created is 64 bit due to
the hypervisor but the dump kernel is 32 bit for maximum compatibility.

It's possibly less likely to be useful in a purely native scenario but I
see no reason to disallow it.

Signed-off-by: Ian Campbell <ian.campbell@xensource.com>
Acked-by: Vivek Goyal <vgoyal@in.ibm.com>
Cc: Horms <horms@verge.net.au>
Cc: Magnus Damm <magnus.damm@gmail.com>
Cc: "Eric W. Biederman" <ebiederm@xmission.com>
Cc: Andi Kleen <ak@suse.de>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>

Index: head-2007-04-27/fs/proc/vmcore.c
===================================================================
--- head-2007-04-27.orig/fs/proc/vmcore.c	2007-04-26 05:08:32.000000000 +0200
+++ head-2007-04-27/fs/proc/vmcore.c	2007-04-27 12:36:55.000000000 +0200
@@ -514,7 +514,7 @@ static int __init parse_crash_elf64_head
 	/* Do some basic Verification. */
 	if (memcmp(ehdr.e_ident, ELFMAG, SELFMAG) != 0 ||
 		(ehdr.e_type != ET_CORE) ||
-		!elf_check_arch(&ehdr) ||
+		!vmcore_elf_check_arch(&ehdr) ||
 		ehdr.e_ident[EI_CLASS] != ELFCLASS64 ||
 		ehdr.e_ident[EI_VERSION] != EV_CURRENT ||
 		ehdr.e_version != EV_CURRENT ||
Index: head-2007-04-27/include/asm-i386/kexec.h
===================================================================
--- head-2007-04-27.orig/include/asm-i386/kexec.h	2007-04-27 12:34:30.000000000 +0200
+++ head-2007-04-27/include/asm-i386/kexec.h	2007-04-27 12:36:55.000000000 +0200
@@ -47,6 +47,9 @@
 /* The native architecture */
 #define KEXEC_ARCH KEXEC_ARCH_386
 
+/* We can also handle crash dumps from 64 bit kernel. */
+#define vmcore_elf_check_arch_cross(x) ((x)->e_machine == EM_X86_64)
+
 #define MAX_NOTE_BYTES 1024
 
 /* CPU does not save ss and esp on stack if execution is already
Index: head-2007-04-27/include/linux/crash_dump.h
===================================================================
--- head-2007-04-27.orig/include/linux/crash_dump.h	2007-04-26 05:08:32.000000000 +0200
+++ head-2007-04-27/include/linux/crash_dump.h	2007-04-27 12:36:55.000000000 +0200
@@ -14,5 +14,13 @@ extern ssize_t copy_oldmem_page(unsigned
 extern const struct file_operations proc_vmcore_operations;
 extern struct proc_dir_entry *proc_vmcore;
 
+/* Architecture code defines this if there are other possible ELF
+ * machine types, e.g. on bi-arch capable hardware. */
+#ifndef vmcore_elf_check_arch_cross
+#define vmcore_elf_check_arch_cross(x) 0
+#endif
+
+#define vmcore_elf_check_arch(x) (elf_check_arch(x) || vmcore_elf_check_arch_cross(x))
+
 #endif /* CONFIG_CRASH_DUMP */
 #endif /* LINUX_CRASHDUMP_H */
